{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MaveDB Mapping Analysis\n",
    "\n",
    "This notebook demonstrates how data from score sets in [MaveDB](https://mavedb.org/) can be mapped to human reference sequences and represented using the [Variation Representation Specification (VRS)](https://vrs.ga4gh.org/) of the [Global Alliance for Genomics and Health (GA4GH)](https://www.ga4gh.org/), as described in \"Mapping MAVE data for use in human genomics applications\" (Arbesfeld et al). \n",
    "\n",
    "Each step of the mapping workflow is demonstrated under the relevant header and accompanied by example data pulled from MaveDB scoreset `urn:mavedb:00000041-a-1`. After each step, data is saved to a local [pickle](https://docs.python.org/3/library/pickle.html) checkpoint file for easy use in later steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "First, initialize environment parameters to enable access to required resources:\n",
    "\n",
    "* Universal Transcript Archive (UTA): see [README](https://github.com/biocommons/uta?tab=readme-ov-file#installing-uta-locally) for setup instructions. Users with access to Docker on their local devices can use the available Docker image; otherwise, start a relatively recent (version 14+) PostgreSQL instance and add data from the `20210129b` database dump.\n",
    "* SeqRepo: see [README](https://github.com/biocommons/biocommons.seqrepo?tab=readme-ov-file#requirements) for setup instructions. Experiments here were run using the `2024-02-20` snapshot.\n",
    "  * Note that `dcd_map` requires writing to SeqRepo's sequence databases. This means the user must have write permissions on the data directory. See [here](https://github.com/biocommons/biocommons.seqrepo/blob/main/docs/store.rst) for more information if using locally-available data.\n",
    "* Gene Normalizer: see [documentation](https://gene-normalizer.readthedocs.io/0.3.0-dev1/install.html) for installation instructions\n",
    "  * This notebook was run using Gene Normalizer PostgreSQL data checkpointed from 2024-05-29. To sync local data against this snapshot, follow [instructions for PostgreSQL setup](https://gene-normalizer.readthedocs.io/0.3.0-dev1/managing_data/postgresql.html#local-setup) and then use the `gene_norm_update_remote` command:\n",
    "\n",
    "```shell\n",
    "$ gene_norm_update_remote --data_url=\"https://vicc-normalizers.s3.us-east-2.amazonaws.com/gene_normalization/postgresql/gene_norm_20240529154335.sql.tar.gz\"\n",
    "```\n",
    "\n",
    "* blat: Must be available on the local PATH and executable by the user. Otherwise, its location can be set manually with the `BLAT_BIN_PATH` environment variable. See the [UCSC Genome Browser FAQ](https://genome.ucsc.edu/FAQ/FAQblat.html#blat3) for download instructions. For our experiments, we placed the binary in the same directory as these notebooks.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from os import environ\n",
    "from pathlib import Path\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# set external resources. configure based on location of available data.\n",
    "environ[\"GENE_NORM_DB_URL\"] = \"postgresql://postgres@localhost:5432/gene_normalizer\"\n",
    "environ[\"UTA_DB_URL\"] = \"postgresql://uta_admin:uta@localhost:5432/uta/uta_20210129b\"\n",
    "environ[\"BLAT_BIN_PATH\"] = str(Path(\"blat\").absolute())\n",
    "environ[\"SEQREPO_ROOT_DIR\"] = \"/usr/local/share/seqrepo/2024-02-20\" \n",
    "environ[\"DCD_MAPPING_RESOURCES_DIR\"] = str(Path(\"./mavedb_files\").absolute())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Output Directory\n",
    "\n",
    "Output from this notebook will be stored in a directory named `analysis_files`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_files_dir = Path(\"analysis_files\")\n",
    "analysis_files_dir.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get experiment data from MaveDB\n",
    "\n",
    "Get metadata for the examined MaveDB score sets (209 in total). Each captures the following:\n",
    "\n",
    "* `urn`: The score set identifier\n",
    "* `target_gene_name`: The listed target for the score set (e.g. Src catalytic domain, CXCR4)\n",
    "* `target_sequence`: The target sequence for the score set\n",
    "* `target_sequence_type`: Is the target sequence a DNA or protein sequence\n",
    "* `target_uniprot_ref`: The Uniprot ID associated with the score set, if available\n",
    "* `target_gene_category`: The target type associated with the score set (e.g. Regulatory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "scoresets_input = Path(\"experiment_scoresets.txt\")\n",
    "with scoresets_input.open() as f:\n",
    "    scoresets = [scoreset.strip() for scoreset in f.readlines()]\n",
    "example_scoreset = \"urn:mavedb:00000041-a-1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cf1b15fc59d74bec93f742a6fa4d59a3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/209 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from dcd_mapping.mavedb_data import get_scoreset_metadata\n",
    "\n",
    "metadata = {}\n",
    "for scoreset in tqdm(scoresets):\n",
    "    metadata[scoreset] = get_scoreset_metadata(scoreset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'urn': 'urn:mavedb:00000041-a-1',\n",
       " 'target_gene_name': 'Src catalytic domain',\n",
       " 'target_gene_category': <TargetType.PROTEIN_CODING: 'Protein coding'>,\n",
       " 'target_sequence': 'CTGCGGCTGGAGGTCAAGCTGGGCCAGGGCTGCTTTGGCGAGGTGTGGATGGGGACCTGGAACGGTACCACCAGGGTGGCCATCAAAACCCTGAAGCCTGGCACGATGTCTCCAGAGGCCTTCCTGCAGGAGGCCCAGGTCATGAAGAAGCTGAGGCATGAGAAGCTGGTGCAGTTGTATGCTGTGGTTTCAGAGGAGCCCATTTACATCGTCACGGAGTACATGAGCAAGGGGAGTTTGCTGGACTTTCTCAAGGGGGAGACAGGCAAGTACCTGCGGCTGCCTCAGCTGGTGGACATGGCTGCTCAGATCGCCTCAGGCATGGCGTACGTGGAGCGGATGAACTACGTCCACCGGGACCTTCGTGCAGCCAACATCCTGGTGGGAGAGAACCTGGTGTGCAAAGTGGCCGACTTTGGGCTGGCTCGGCTCATTGAAGACAATGAGTACACGGCGCGGCAAGGTGCCAAATTCCCCATCAAGTGGACGGCTCCAGAAGCTGCCCTCTATGGCCGCTTCACCATCAAGTCGGACGTGTGGTCCTTCGGGATCCTGCTGACTGAGCTCACCACAAAGGGACGGGTGCCCTACCCTGGGATGGTGAACCGCGAGGTGCTGGACCAGGTGGAGCGGGGCTACCGGATGCCCTGCCCGCCGGAGTGTCCCGAGTCCCTGCACGACCTCATGTGCCAGTGCTGGCGGAAGGAGCCTGAGGAGCGGCCCACCTTCGAGTACCTGCAGGCCTTCCTG',\n",
       " 'target_sequence_type': <TargetSequenceType.DNA: 'dna'>,\n",
       " 'target_uniprot_ref': {'id': 'uniprot:P12931', 'offset': 269}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[example_scoreset].model_dump()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, get corresponding experiment scores from MaveDB. Mirroring information provided by `/scores` API endpoint, provides the following data for each score in a score set:\n",
    "\n",
    "* `hgvs_pro`: variant description with respect to the amino acid target sequence\n",
    "* `hgvs_nt`: variant description with respect to the nucleotide target sequence\n",
    "* `score`: raw reported score\n",
    "* `accession`: accession identifier for the specific experiment, e.g. `urn:mavedb:00000041-a-1#548`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0514dfe531f7471b91f051ae1602595c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/209 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from dcd_mapping.mavedb_data import get_scoreset_records\n",
    "\n",
    "scores = {}\n",
    "for urn in tqdm(scoresets):\n",
    "    try:\n",
    "        scores[urn] = get_scoreset_records(urn)\n",
    "    except:\n",
    "        print(urn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each score \"row\" consists of nucleotide and/or protein [MAVE-HGVS expressions](https://www.mavedb.org/docs/mavehgvs/spec.html), a score, and an identifier:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'hgvs_pro': 'p.Tyr170Gly',\n",
       " 'hgvs_nt': 'NA',\n",
       " 'score': '0.753146338',\n",
       " 'accession': 'urn:mavedb:00000041-a-1#36'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scores[example_scoreset][0].model_dump()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: MaveDB Metadata to BLAT Alignment Data\n",
    "\n",
    "During this step, the target sequence for each score set is run through BLAT, allowing for genomic coordinates to be linked with the target sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate BLAT Output for each Score Set\n",
    "\n",
    "Generate BLAT alignment output for each examined score set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b9d1c6b8e3e8499884ad176721a0cd25",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/209 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383583v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270750v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_GL000224v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr12_GL383550v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr9_GL383539v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr2_KI270893v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr9_KI270719v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_KI270935v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270890v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270875v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr9_GL383541v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270817v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270744v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr21_GL383579v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270932v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr10_GL383545v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr7_KI270805v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_KI270790v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr5_GL949742v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_KI270778v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270866v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270746v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr9_KI270823v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270926v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270914v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_KI270849v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270832v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270831v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270710v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270706v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_GL383565v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_GL383555v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270826v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270816v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr7_KI270809v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_KI270786v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_KI270936v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_KI270777v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr20_KI270870v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr20_GL383577v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270931v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270885v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr18_KI270864v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_GL383566v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_KI270852v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr14_GL000225v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr10_KI270824v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270742v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_GL000213v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270822v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_JH636055v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_GL383526v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_GL000221v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270878v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270876v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270765v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270930v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270929v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270922v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270921v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270887v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL383573v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL000209v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr18_KI270863v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr16_KI270728v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_KI270851v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr10_KI270825v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr5_KI270897v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_KI270788v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr3_KI270781v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr21_GL383581v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr21_GL383580v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr21_GL383578v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270763v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_KI270848v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr14_KI270844v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrX_KI270880v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270815v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_KI270800v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000256v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000253v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000250v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr5_GL339449v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_KI270789v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270879v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270766v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270759v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_GL383518v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270938v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270883v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949753v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949752v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949751v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949750v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949748v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949747v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949746v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr18_KI270911v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_KI270908v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_KI270862v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_GL383563v3_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_GL000258v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_KI270905v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr14_KI270847v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr12_KI270834v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr14_KI270845v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_KI270798v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_KB021644v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr10_GL383546v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270748v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr13_KI270841v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr2_KI270770v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr14_GL000009v2_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr7_KI270899v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr12_GL383552v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_KI270785v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_GL000216v2\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr2_KI270774v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr2_KI270768v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr13_KI270843v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr2_KI270769v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_KI270925v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr14_GL000194v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270814v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270927v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270811v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr4_GL000257v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270708v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270916v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270891v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270888v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270886v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270882v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_KI270859v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr12_KI270835v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270903v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270721v1_random\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270442v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr8_KI270818v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr7_KI270806v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr7_KI270803v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_KI270761v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270923v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL949749v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_GL383574v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr18_GL383567v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_KI270909v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr16_KI270856v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_GL383554v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr12_KI270904v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chrUn_KI270751v1\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr7_KI270808v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr18_GL383568v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr17_KI270861v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270832v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270832v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_GL383519v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270832v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383583v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr1_GL383519v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr11_KI270832v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383583v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000256v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000255v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000254v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000252v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr6_GL000251v2_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr5_GL949742v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270915v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270891v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr19_KI270889v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr15_KI270905v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KI270928v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_KB663609v1_alt\n",
      "SeqRepo unable to get translated identifiers for GRCh37:chr22_GL383582v2_alt\n"
     ]
    }
   ],
   "source": [
    "from dcd_mapping.align import AlignmentError, align\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata\n",
    "\n",
    "align_results = {}\n",
    "failed_alignment_scoresets = []\n",
    "\n",
    "for scoreset, meta in tqdm(metadata.items()):\n",
    "    try:\n",
    "        align_results[scoreset] = align(meta, silent=True)\n",
    "    except AlignmentError:\n",
    "        failed_alignment_scoresets.append(scoreset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*(These SeqRepo warnings are expected -- part of the alignment step involves matching accessions for blat candidate chromosomes against an expected accession given scoreset annotations, but some blat candidates come from alternative accessions that we don't retain data for)*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "During our experiments, we found that one scoreset, `urn:mavedb:00000105-a-1`, fails to return a BLAT hit against the reference genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['urn:mavedb:00000105-a-1']\n"
     ]
    }
   ],
   "source": [
    "print(failed_alignment_scoresets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result of the alignment phase is a structured description of the best BLAT result for the input sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chrom': 'chr20',\n",
       " 'strand': <Strand.POSITIVE: 1>,\n",
       " 'coverage': 100.0,\n",
       " 'ident_pct': 99.86666666666666,\n",
       " 'query_range': {'start': 0, 'end': 750},\n",
       " 'query_subranges': [{'start': 0, 'end': 52},\n",
       "  {'start': 52, 'end': 232},\n",
       "  {'start': 232, 'end': 309},\n",
       "  {'start': 309, 'end': 463},\n",
       "  {'start': 463, 'end': 595},\n",
       "  {'start': 595, 'end': 750}],\n",
       " 'hit_range': {'start': 37397802, 'end': 37403325},\n",
       " 'hit_subranges': [{'start': 37397802, 'end': 37397854},\n",
       "  {'start': 37400114, 'end': 37400294},\n",
       "  {'start': 37401601, 'end': 37401678},\n",
       "  {'start': 37402434, 'end': 37402588},\n",
       "  {'start': 37402748, 'end': 37402880},\n",
       "  {'start': 37403170, 'end': 37403325}]}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "align_results[example_scoreset].model_dump()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save BLAT Output\n",
    "\n",
    "Save a checkpoint for the BLAT results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "mave_blat_to_save = {}\n",
    "for scoreset, result in align_results.items():\n",
    "    mave_blat_to_save[scoreset] = result.model_dump(exclude_none=True)\n",
    "with (analysis_files_dir / \"mave_blat_output.pickle\").open(\"wb\") as fn:\n",
    "    pickle.dump(mave_blat_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "del mave_blat_to_save"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Transcript and Offset Selection for MaveDB Score Sets\n",
    "\n",
    "In this phase, a human transcript is chosen for each protein-coding score set, and an offset is computed when the target sequence does not occur at the start of the human reference sequence. For regulatory/other non-coding score sets, a transcript is not chosen and the chromosomal sequence is selected as the reference sequence. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load BLAT output\n",
    "\n",
    "Load checkpointed BLAT output for the examined MaveDB score sets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult\n",
    "\n",
    "with (analysis_files_dir / \"mave_blat_output.pickle\").open(\"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for scoreset in scoresets:\n",
    "    align_result = mave_blat_temp.get(scoreset)\n",
    "    if align_result:\n",
    "        align_results[scoreset] = AlignmentResult(**align_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Transcript Mappings File\n",
    "\n",
    "Generate a transcript mapping for each relevant score set containing the following data:\n",
    "\n",
    "* `nm`: A RefSeq transcript accession\n",
    "* `np`: A RefSeq protein sequence accession\n",
    "* `start`: An integer containing the offset for the target sequence with the respect to the selected human reference sequence\n",
    "* `transcript_mode`: The set of [MANE annotations](https://www.ncbi.nlm.nih.gov/refseq/MANE/) in which the selected transcript is included. See the [CoolSeqTool docs](https://coolseqtool.readthedocs.io/0.4.0-dev3/transcript_selection.html#representative-transcript-priority) for additional information\n",
    "* `sequence`: The translated protein reference sequence\n",
    "* `is_full_match`: sequence is a complete match for target sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "34ea0bf6b17a4e4ea70e44ce661791eb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/209 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting transcript start = 0 -- there is discordance between actual and expected amino acid locations in experiments 47 and 48\n",
      "urn:mavedb:00000058-a-1 describes the starting residue as Asp2, but the starting residue is D -- manually reducing offset by 1 to reflect start of Met1.\n",
      "Setting transcript start = 0 -- there is discordance between actual and expected amino acid locations in experiments 47 and 48\n",
      "Experiment 53's target sequence is missing start residue E -- manually reducing offset by 1\n",
      "Setting transcript start = 0 -- there is discordance between actual and expected amino acid locations in experiments 47 and 48\n",
      "Experiment 53's target sequence is missing start residue E -- manually reducing offset by 1\n",
      "Setting transcript start = 0 -- there is discordance between actual and expected amino acid locations in experiments 47 and 48\n",
      "Setting transcript start = 0 -- there is discordance between actual and expected amino acid locations in experiments 47 and 48\n",
      "Setting transcript start = 0 -- there is discordance between actual and expected amino acid locations in experiments 47 and 48\n"
     ]
    }
   ],
   "source": [
    "import asyncio\n",
    "\n",
    "import nest_asyncio\n",
    "\n",
    "from dcd_mapping.transcripts import TxSelectError, select_transcript\n",
    "\n",
    "nest_asyncio.apply()\n",
    "failed_tx_select_scoresets = [] \n",
    "tx_selection = {}\n",
    "for ss in tqdm(scoresets):\n",
    "    if ss in align_results:\n",
    "        try:\n",
    "            tx_selection[ss] = asyncio.run(\n",
    "                select_transcript(\n",
    "                    metadata[ss],\n",
    "                    scores[ss],\n",
    "                    align_results[ss],\n",
    "                )\n",
    "            )\n",
    "        except TxSelectError:\n",
    "            failed_tx_select_scoresets.append(ss)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Warnings here reflect manual adjustments needed to correct for reporting errors*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This phase should be completed without encountering any new errors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "failed_tx_select_scoresets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transcript selection and offset data is stored for each scoreset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nm': 'NM_198291.3',\n",
       " 'np': 'NP_938033.1',\n",
       " 'start': 269,\n",
       " 'is_full_match': True,\n",
       " 'transcript_mode': <TranscriptPriority.MANE_SELECT: 'mane_select'>,\n",
       " 'sequence': 'LRLEVKLGQGCFGEVWMGTWNGTTRVAIKTLKPGTMSPEAFLQEAQVMKKLRHEKLVQLYAVVSEEPIYIVTEYMSKGSLLDFLKGETGKYLRLPQLVDMAAQIASGMAYVERMNYVHRDLRAANILVGENLVCKVADFGLARLIEDNEYTARQGAKFPIKWTAPEAALYGRFTIKSDVWSFGILLTELTTKGRVPYPGMVNREVLDQVERGYRMPCPPECPESLHDLMCQCWRKEPEERPTFEYLQAFL'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_selection[example_scoreset].model_dump()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save Transcript Mappings Output\n",
    "\n",
    "Save a checkpoint for the `transcript_mappings` data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "transcript_mappings_to_save = {}\n",
    "for ss in tx_selection:\n",
    "    if tx_selection[ss]:\n",
    "        transcript_mappings_to_save[ss] = tx_selection[ss].model_dump(exclude_none=True)\n",
    "with (analysis_files_dir / \"transcript_mappings.pickle\").open(\"wb\") as fn:\n",
    "    pickle.dump(transcript_mappings_to_save, fn, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "del transcript_mappings_to_save"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Mapping MAVE Variants using the GA4GH Variation Representation Specification (VRS)\n",
    "\n",
    "During this phase, MAVE variants are supplied to VRS, generating a pre-mapped and post-mapped computable representation for each variant. The functional effect score for each variant pair and the associated MaveDB ID are also stored in separate dictionaries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Alignment and Transcript Selection Data\n",
    "\n",
    "Load checkpointed alignment and transcript selection data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "from dcd_mapping.schemas import AlignmentResult, TxSelectResult\n",
    "\n",
    "with (analysis_files_dir / \"mave_blat_output.pickle\").open(\"rb\") as fn:\n",
    "    mave_blat_temp = pickle.load(fn)\n",
    "align_results = {}\n",
    "for ss in mave_blat_temp:\n",
    "    align_results[ss] = AlignmentResult(**mave_blat_temp[ss])\n",
    "del mave_blat_temp\n",
    "\n",
    "with (analysis_files_dir / \"transcript_mappings.pickle\").open(\"rb\") as fn:\n",
    "    transcript_mappings_temp = pickle.load(fn)\n",
    "tx_selection = {}\n",
    "for ss in transcript_mappings_temp:\n",
    "    tx_selection[ss] = TxSelectResult(**transcript_mappings_temp[ss])\n",
    "del transcript_mappings_temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert MaveDB Variants to VRS Alleles\n",
    "\n",
    "Convert MaveDB variants to VRS objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1ba8986d7a7842d58cde2ae042998cd9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/208 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "OperationalError",
     "evalue": "attempt to write a readonly database",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m mave_vrs_mappings \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m ss \u001b[38;5;129;01min\u001b[39;00m tqdm(align_results):\n\u001b[0;32m----> 8\u001b[0m     mave_vrs_mappings[ss] \u001b[38;5;241m=\u001b[39m \u001b[43mvrs_map\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m[\u001b[49m\u001b[43mss\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m        \u001b[49m\u001b[43malign_results\u001b[49m\u001b[43m[\u001b[49m\u001b[43mss\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscores\u001b[49m\u001b[43m[\u001b[49m\u001b[43mss\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtx_selection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mss\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/dcd-mapping/src/dcd_mapping/vrs_map.py:491\u001b[0m, in \u001b[0;36mvrs_map\u001b[0;34m(metadata, align_result, records, transcript, silent)\u001b[0m\n\u001b[1;32m    488\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    490\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m metadata\u001b[38;5;241m.\u001b[39mtarget_gene_category \u001b[38;5;241m==\u001b[39m TargetType\u001b[38;5;241m.\u001b[39mPROTEIN_CODING \u001b[38;5;129;01mand\u001b[39;00m transcript:\n\u001b[0;32m--> 491\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_map_protein_coding\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrecords\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtranscript\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43malign_result\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _map_regulatory_noncoding(\n\u001b[1;32m    498\u001b[0m     metadata,\n\u001b[1;32m    499\u001b[0m     records,\n\u001b[1;32m    500\u001b[0m     align_result,\n\u001b[1;32m    501\u001b[0m )\n",
      "File \u001b[0;32m~/code/dcd-mapping/src/dcd_mapping/vrs_map.py:232\u001b[0m, in \u001b[0;36m_map_protein_coding\u001b[0;34m(metadata, records, transcript, align_result)\u001b[0m\n\u001b[1;32m    228\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m metadata\u001b[38;5;241m.\u001b[39mtarget_sequence_type \u001b[38;5;241m==\u001b[39m TargetSequenceType\u001b[38;5;241m.\u001b[39mDNA:\n\u001b[1;32m    229\u001b[0m     sequence \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\n\u001b[1;32m    230\u001b[0m         Seq(metadata\u001b[38;5;241m.\u001b[39mtarget_sequence)\u001b[38;5;241m.\u001b[39mtranslate(table\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m1\u001b[39m\u001b[38;5;124m\"\u001b[39m, stop_symbol\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    231\u001b[0m     )\n\u001b[0;32m--> 232\u001b[0m     psequence_id \u001b[38;5;241m=\u001b[39m \u001b[43mstore_sequence\u001b[49m\u001b[43m(\u001b[49m\u001b[43msequence\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    233\u001b[0m     gsequence_id \u001b[38;5;241m=\u001b[39m store_sequence(metadata\u001b[38;5;241m.\u001b[39mtarget_sequence)\n\u001b[1;32m    234\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/code/dcd-mapping/src/dcd_mapping/vrs_map.py:197\u001b[0m, in \u001b[0;36mstore_sequence\u001b[0;34m(sequence)\u001b[0m\n\u001b[1;32m    195\u001b[0m alias_dict_list \u001b[38;5;241m=\u001b[39m [{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnamespace\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mga4gh\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124malias\u001b[39m\u001b[38;5;124m\"\u001b[39m: sequence_id}]\n\u001b[1;32m    196\u001b[0m sr \u001b[38;5;241m=\u001b[39m get_seqrepo()\n\u001b[0;32m--> 197\u001b[0m \u001b[43msr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstore\u001b[49m\u001b[43m(\u001b[49m\u001b[43msequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43malias_dict_list\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m sequence_id\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.2/lib/python3.12/site-packages/biocommons/seqrepo/seqrepo.py:234\u001b[0m, in \u001b[0;36mSeqRepo.store\u001b[0;34m(self, seq, nsaliases)\u001b[0m\n\u001b[1;32m    232\u001b[0m     _logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPrecommit for large sequence\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    233\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcommit()\n\u001b[0;32m--> 234\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msequences\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstore\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseq_id\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mseq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    235\u001b[0m n_seqs_added \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    236\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pending_sequences \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.2/lib/python3.12/site-packages/biocommons/seqrepo/fastadir/fastadir.py:207\u001b[0m, in \u001b[0;36mFastaDir.store\u001b[0;34m(self, seq_id, seq)\u001b[0m\n\u001b[1;32m    205\u001b[0m alpha \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28msorted\u001b[39m(\u001b[38;5;28mset\u001b[39m(seq)))\n\u001b[1;32m    206\u001b[0m cursor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_db\u001b[38;5;241m.\u001b[39mcursor()\n\u001b[0;32m--> 207\u001b[0m \u001b[43mcursor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    208\u001b[0m \u001b[38;5;250;43m    \u001b[39;49m\u001b[38;5;124;43;03m\"\"\"insert into seqinfo (seq_id, len, alpha, relpath)\u001b[39;49;00m\n\u001b[1;32m    209\u001b[0m \u001b[38;5;124;43;03m                 values (?, ?, ?,?)\"\"\"\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    210\u001b[0m \u001b[43m    \u001b[49m\u001b[43m(\u001b[49m\u001b[43mseq_id\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mseq\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43malpha\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_writing\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrelpath\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    211\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    212\u001b[0m cursor\u001b[38;5;241m.\u001b[39mclose()\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m seq_id\n",
      "\u001b[0;31mOperationalError\u001b[0m: attempt to write a readonly database"
     ]
    }
   ],
   "source": [
    "from dcd_mapping.lookup import get_seqrepo\n",
    "from dcd_mapping.vrs_map import vrs_map\n",
    "from dcd_mapping.mavedb_data import get_scoreset_metadata, get_scoreset_records\n",
    "\n",
    "mave_vrs_mappings = {}\n",
    "\n",
    "for ss in tqdm(align_results):\n",
    "    mave_vrs_mappings[ss] = vrs_map(\n",
    "        metadata[ss],\n",
    "        align_results[ss],\n",
    "        scores[ss],\n",
    "        tx_selection.get(ss)\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each score in a MaveDB scoreset, VRS objects are generated from both the original (pre-mapped) MAVE variation descriptions as well as the variations that have been mapped to reference transcripts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mave_vrs_mappings[example_scoreset][0].model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS Mappings Dictionary\n",
    "\n",
    "Save a checkpoint of the VRS mappings dictionary to `analysis_files`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "tmp_mave_vrs_mappings = {}\n",
    "for ss, mappings in mave_vrs_mappings.items():\n",
    "    if mappings:\n",
    "        tmp_mave_vrs_mappings[ss] = [m.model_dump(exclude_none=True, exclude_unset=True) for m in mappings]\n",
    "    else:\n",
    "        tmp_mave_vrs_mappings[ss] = mappings\n",
    "with (analysis_files_dir / \"mave_vrs_mappings.pickle\").open(\"wb\") as fn:\n",
    "    pickle.dump(tmp_mave_vrs_mappings, fn, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "del tmp_mave_vrs_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Annotations\n",
    "\n",
    "Finally, annotate MaveDB scoreset metadata with the pre- and post-mapped VRS objects, as well as two additional data points:\n",
    "\n",
    "1. `vrs_ref_allele_seq`: The sequence between the start and end positions indicated in the variant\n",
    "2. `hgvs`: An HGVS string describing the variant (only included for post-mapped variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcd_mapping.annotate import annotate\n",
    "\n",
    "annotated_vrs_mappings = {}\n",
    "for urn, mapping in tqdm(mave_vrs_mappings.items()):\n",
    "    if mapping:\n",
    "        annotated_vrs_mappings[urn] =  annotate(mapping, tx_selection.get(urn), metadata[urn])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final product provided to our reported integration projects includes VRS 1.3-compliant alleles:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotated_vrs_mappings[example_scoreset][0].model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save VRS mappings output in JSON files\n",
    "\n",
    "Run the cells below to save the VRS mappings output in a JSON file in `analysis_files/mappings`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcd_mapping.annotate import save_mapped_output_json\n",
    "\n",
    "mappings_dir = analysis_files_dir / \"mappings\"\n",
    "mappings_dir.mkdir(exist_ok=True)\n",
    "\n",
    "for urn, mappings in tqdm(annotated_vrs_mappings.items()):\n",
    "    output_file = mappings_dir / f\"{urn}_mappings.json\"\n",
    "    save_mapped_output_json(\n",
    "        urn,\n",
    "        mappings,\n",
    "        align_results[urn],\n",
    "        tx_selection.get(urn),\n",
    "        include_vrs_2=False,\n",
    "        output_path=output_file\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dcd-mapping",
   "language": "python",
   "name": "dcd-mapping"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
